File: 0670-maximum-swap/0670-maximum-swap.py
class Solution:
    def maximumSwap(self, num: int) -> int:
        s = list(str(num))
        
        # Loop through each digit in the number
        for i in range(len(s)):
            # Try to find a larger digit that appears later in the string
            for d in range(9, int(s[i]), -1):
                j = ''.join(s).rfind(str(d))
                # If such a digit is found and it's after the current position
                if j > i:
                    # Swap the digits
                    s[i], s[j] = s[j], s[i]
                    # Return the result after converting the list back to integer
                    return int(''.join(s))
        
        return num  # If no swap is made, return the original number

File: 0670-maximum-swap/0670-maximum-swap.cpp
class Solution {
public:
    int maximumSwap(int num) {
        string s = to_string(num);
        for (int i = 0, j = -1; i < s.size(); ++i)
            for (int d = 9; d > s[i] - '0'; --d)
                if ((j = s.find_last_of('0' + d)) > i) {
                    swap(s[i], s[j]);
                    return stoi(s);
                }
        return num;
    }
};
